pjrt-plugin/iree/integrations/pjrt/common/platform.cc

```diff
@@ -50,6 +53,134 @@ void Logger::error(std::string_view message) {
   std::cerr << "[IREE-PJRT] ERROR: " << message << std::endl;
 }
 
+//===----------------------------------------------------------------------===//
+// ArtifactDumper
+//===----------------------------------------------------------------------===//
+
+ArtifactDumper::~ArtifactDumper() = default;
+
+std::unique_ptr<ArtifactDumper::Transaction>
+ArtifactDumper::CreateTransaction() {
+  return nullptr;
+}
+
+std::string ArtifactDumper::DebugString() { return std::string("disabled"); }
+
+//===----------------------------------------------------------------------===//
+// FilesArtifactDumper
+//===----------------------------------------------------------------------===//
+
+class FilesArtifactDumper::FilesTransaction
+    : public ArtifactDumper::Transaction {
+ public:
+  FilesTransaction(Logger& logger, std::filesystem::path base_path,
+                   int64_t transaction_id, bool retain_all)
+      : logger_(logger),
+        base_path_(std::move(base_path)),
+        transaction_id_(transaction_id),
+        retain_all_(retain_all) {}
+  ~FilesTransaction() { Retain(); }
+
+  void WriteArtifact(std::string_view label, std::string_view extension,
+                     int index, std::string_view contents) override {
+    std::string basename = std::to_string(transaction_id_);
+    basename.append("-");
+    basename.append(label);
+    if (index >= 0) {
+      basename.append(std::to_string(index));
+    }
+    basename.append(".");
+    basename.append(extension);
+
+    auto file_path = base_path_ / basename;
+    std::ofstream fout;
+    fout.open(file_path, std::ofstream::out | std::ofstream::trunc |
+                             std::ofstream::binary);
+    fout.write(contents.data(), contents.size());
+    fout.close();
+
+    written_paths_.push_back(file_path);
+
+    if (!fout.good()) {
+      std::string message("I/O error dumping artifact: ");
+      message.append(file_path);
+      logger_.error(message);
+    }
+  }
+
+  void Retain() override {
+    if (written_paths_.empty()) return;
+
+    std::string message;
+    message.append("Retained artifacts in: ");
+    message.append(base_path_);
+    for (auto& p : written_paths_) {
+      message.append("\n  ");
+      message.append(p.filename());
+    }
+    logger_.debug(message);
+
+    written_paths_.clear();
+  }
+
+  void Cancel() override {
+    if (retain_all_) {
+      Retain();
+      return;
+    }
+
+    for (auto& p : written_paths_) {
+      std::error_code ec;
+      std::filesystem::remove(p, ec);
+      if (ec) {
+        // Only carp as a debug message since there are legitimate reasons
+        // this can happen depending on what is going on at the system level.
+        std::string message("Error removing artifact: ");
+        message.append(p);
+        logger_.debug(message);
+      }
+    }
+
+    written_paths_.clear();
+  }
+
+ private:
+  Logger& logger_;
```

Comment:
I think it would be smaller class layout if reorganized, but that's not critical.

Reply:
I can usually spot those but missing this time. Looks like a bunch of pointer sized things, followed by bool... which is the right way, right?